docs/design/features/source-generator-pinvokes.md

```diff
@@ -0,0 +1,232 @@
+
+2) The runtime attempts to find a binary with the name supplied in `DllImportAttribute`.
+
+    * Discovery of the target binary is complicated and can be influenced by the [`AssemblyLoadContext`](https://docs.microsoft.com/dotnet/api/system.runtime.loader.assemblyloadcontext) and [`NativeLibrary`](https://docs.microsoft.com/dotnet/api/system.runtime.interopservices.nativelibrary) classes.
+
+3) Once the binary is found and loaded into the runtime, it is queried for the expected export name. The name of the attributed function is used by default but this is configurable by the [`DllImportAttribute.EntryPoint`](https://docs.microsoft.com/dotnet/api/system.runtime.interopservices.dllimportattribute.entrypoint) property.
+
+    * This process is also influenced by additional `DllImportAttribute` properties as well as by the underlying platform. For example, the [Win32 API ANSI/UNICODE convention](https://docs.microsoft.com/windows/win32/intl/conventions-for-function-prototypes) on Windows is respected and a(n) `A`/`W` suffix may be appended to the function name if it is not immediately found.
+
+4) The IL Stub is called like any other .NET method.
+
+5) The IL Stub marshals arguments as appropriate and invokes the export via the `calli` instruction.
+
+6) Once the export returns control to the IL Stub the marshaling logic cleans up and ensures any returned data is marshaled back out to the calling function.
+
+### Source Generator Integration
+
+An example of how the previous P/Invoke snippet could be transformed is below. This example is using the proposed API in this document. The Source Generator currently has a limitation that we will ignore at present - the inability to modify user written code. This limitation is ignored to present an optimal user experience, but will be discussed later in case the current limitation becomes a hard requirement.
+
+`Program.cs` (Pre-generated code)
+
+``` CSharp
+/* A */ [DllImportAttribute("Kernel32.dll")]
+/* D */ [GeneratedDllImportAttribute]
+/* B */ extern static bool QueryPerformanceCounter(out long lpPerformanceCount);
+...
+long count;
+/* C*/ QueryPerformanceCounter(out count);
+```
+
+Observe point (D), the new attribute. This attribute provides an indication to a Source Generator that the following declaration should be updated.
+
+`Program.cs` (In-memory updated code)
+
+``` CSharp
+/* D */ [GeneratedDllImportAttribute]
+/* B */ partial static bool QueryPerformanceCounter(out long lpPerformanceCount);
+...
+long count;
+/* C */ QueryPerformanceCounter(out count);
+```
+
+During the source generation process the `DllImportAttribute` (A) would be removed. The `GeneratedDllImportAttribute` (D) would remain to provide a metadata indication that the function was auto-generated. Also note that the method declaration has been updated and marked `partial`. The Source Generator would then generate the source for this partial method. The invocation (C) remains unchanged.
```

Comment:
> Durring the source generation process the `DllImportAttribute` (A) would be removed.

Source generators do not allow for modification of existing source code.

Reply:
Agreed. This is mentioned in the premise for the associated section.